File: frontend/src/api/matches.ts
import { api } from "@/lib/api";
import type { ArenaStats, Match, MatchActionKind, MatchDifficulty } from "@/types";

export async function playMatch(difficulty: MatchDifficulty): Promise<Match> {
  const { data } = await api.post<Match>("/matches/play", { difficulty });
  return data;
}

export async function actMatch(matchId: number, action: MatchActionKind): Promise<Match> {
  const { data } = await api.post<Match>(`/matches/${matchId}/act`, { action });
  return data;
}

export async function forfeitMatch(matchId: number): Promise<Match> {
  const { data } = await api.post<Match>(`/matches/${matchId}/forfeit`);
  return data;
}

export async function fetchMatchHistory(): Promise<Match[]> {
  const { data } = await api.get<Match[]>("/matches/history");
  return data;
}

export async function fetchArenaStats(): Promise<ArenaStats> {
  const { data } = await api.get<ArenaStats>("/matches/stats");
  return data;
}

File: backend/app/routers/matches.py
from fastapi import APIRouter, Depends
from sqlalchemy.ext.asyncio import AsyncSession

from app.core.dependencies import get_current_user
from app.core.rate_limit import check_rate_limit
from app.database import get_db
from app.models.user import User
from app.schemas.match import ArenaLeaderboardEntry, ArenaStatsOut, MatchActionRequest, MatchOut, StartMatchRequest
from app.services import match_service

router = APIRouter(prefix="/matches", tags=["matches"])


@router.post("/play", response_model=MatchOut)
async def play_match(
    payload: StartMatchRequest, db: AsyncSession = Depends(get_db), user: User = Depends(get_current_user)
):
    check_rate_limit(f"play_match:{user.id}", max_calls=15, window_seconds=60)
    return await match_service.start_match(db, user, payload)


@router.post("/{match_id}/act", response_model=MatchOut)
async def act(
    match_id: int, payload: MatchActionRequest, db: AsyncSession = Depends(get_db), user: User = Depends(get_current_user)
):
    return await match_service.resolve_action(db, user, match_id, payload)


@router.post("/{match_id}/forfeit", response_model=MatchOut)
async def forfeit(match_id: int, db: AsyncSession = Depends(get_db), user: User = Depends(get_current_user)):
    return await match_service.forfeit_match(db, user, match_id)


@router.get("/history", response_model=list[MatchOut])
async def get_match_history(db: AsyncSession = Depends(get_db), user: User = Depends(get_current_user)):
    return await match_service.match_history(db, user)


@router.get("/stats", response_model=ArenaStatsOut)
async def get_arena_stats(db: AsyncSession = Depends(get_db), user: User = Depends(get_current_user)):
    return await match_service.arena_stats(db, user)


@router.get("/leaderboard", response_model=list[ArenaLeaderboardEntry])
async def get_arena_leaderboard(db: AsyncSession = Depends(get_db), _user: User = Depends(get_current_user)):
    return await match_service.arena_leaderboard(db)


@router.get("/{match_id}", response_model=MatchOut)
async def get_match_detail(match_id: int, db: AsyncSession = Depends(get_db), user: User = Depends(get_current_user)):
    return await match_service.get_match(db, user, match_id)
